### DISCUSSION TITLE
Command line tool?

### DISCUSSION THREAD

**Discussion by georgeef** [moneymanagerex/moneymanagerex#6879] [Ideas]
I started working on export/import to/from XLSX, and it is already clear that this will overload MMEX Desktop with at least one more external library. In this context, I made some thoughts which I share below, hoping for your feedback.

The idea is to have a command-line tool for less often needed operations. It should support the same platforms as MMEX Desktop, but without any GUI. I can think of the following operations:

- Database consistency checks.

- Data checks (e.g., similar payee names).

- Split database in two parts (historical and current), at a given date.

- Merge two databases.

- Show the difference between two databases.

- Initialize a new database using info from an old database.

- Export/import to/from XLSX or other formats in the future, which require external libraries.

- Print short reports in plain text, for easy copy-paste.

Advantages of a separate command-line tool:

- It offloads MMEX Desktop from features that are not needed every day, also from the corresponding external libraries.

- It can better focus on database operations, without the complexity of a GUI.

- It can better handle databases generated from different apps, or it can support more historical versions.

- It can be used strictly in read-only mode, therefore it is safer than MMEX Desktop.

- Less release stress. MMEX Desktop can be released even if the command-line tool is not ready for release.

There are also several disadvantages:

- One more tool to maintain and to keep in sync with respect to model changes.

- Most Windows and macOS users don't use (or don't want to use) command-line tools.

- Mobile users cannot use command-line tools (but there is no good alternative, unless all operations are implemented also in mobile apps, which is a lot of effort for little value).

- Most desktop users prefer a single application for all purposes.

- MMEX Desktop can provide a configuration GUI for complex database operations, which offers better and more intuitive control, compared to cryptic command-line options.

- Potential confusion about the compatibility of different versions of command-line tool, desktop, and mobile apps.

@guanlisheng @vomikan @whalley @n-stein 
What are your thoughts?

**Comment by whalley** [comment]
So long as the command line tool(s) are adequately documented I'm supportive of this.

**Comment by vomikan** [comment]
> * Data checks (e.g., similar payee names).

GRM may be used for this purposes https://github.com/moneymanagerex/general-reports/tree/master/packages/Transaction/InvalidTransactions



> * Split database in two parts (historical and current), at a given data.

https://github.com/moneymanagerex/moneymanagerex/issues/2760



> * Merge two databases.

it may be useful to someone in rare cases



> * Show the difference between two databases.

like



> * Initialize a new database using info from an old database.

like

Export/import to/from XLSX or other formats in the future, which require external libraries.

I don't like

> Print short reports in plain text, for easy copy-paste.

like

**Reply by georgeef** [reply]
> > * Split database in two parts (historical and current), at a given ~~data~~ date.
> 
> #2760

#2760 does only a cleanup. this is sufficient for the historical part (remove everything after a given date, as if time stopped there; also remove all recurring transactions). for the current part, all account balances must be calculated at the split date and stored as initial balances, and all transactions before the split date must be removed. at the end there is one new file (the historical part), which is meant to be an archive.

**Reply by georgeef** [reply]
> Export/import to/from XLSX or other formats in the future, which require external libraries.
> 
> I don't like

could you add some explanation why you don't like export/import to XLSX?  or just an export to XLSX, which cannot corrupt the database. this feature has been requested by users, and the CSV format has severe limitations and many reported problems.

**Reply by vomikan** [reply]
I'm not a fan of the xlsx format in general. It's proprietary and good for presenting data, but not for data exchange.
And it seems we already have report export in this format.

**Reply by georgeef** [reply]
it can also be ODS, but i cannot find a good driver for it. LibreOffice supports both without any problem. XLSX is essentially open format now. i use only text and numbers (see #6837 for details).

you are correct, CSV is more of an interchange format. the main problem is that in CSV you can export only one table. so to export the whole database you need to export many tables in many files. XLSX/ODS combines all these tables in one file, from where you can export to CSV if needed. most users would like to view/process data in XLSX/ODS, with CSV they can only import to another application.

**Reply by vomikan** [reply]
I tried to make JSON export and import. Partly even implemented it. I like this more as a backup tool, or as a data set. But not for everyone.

**Reply by georgeef** [reply]
ok, i put #6837 on hold.

**Comment by vomikan** [comment]
I think that a separate application is bad. It is best to do this inside the desktop version.

**Reply by guanlisheng** [reply]
Yup, I second to have a single application. 

With great architecture, most of these would be implemented as a plugin, just like what we have done in GRM.

in terms of export/import format,  I think json should be considered. In the mmex4ios, which was just started, json export is build-in function for all entity. 
her you go a simplified example

```

{
  "status" : "Open",
  "currencyId" : 3,
  "name" : "Current2",
  "currency" : {
    "decimalPoint" : ".",
    "type" : "Fiat",
    "scale" : 100,
    "unitName" : "",
    "groupSeparator" : " ",
    "name" : "UK Pound",
    "centName" : "",
    "id" : 3,
    "symbol" : "GBP",
    "suffixSymbol" : "",
    "prefixSymbol" : "£"
  },
  "type" : "Checking",
  "favoriteAcct" : "TRUE",
  "notes" : "",
  "id" : 3
}

```

**Reply by georgeef** [reply]
i stopped working on export/import to XLSX (i have my tool for personal use), but anyway i would like to comment that one format does not exclude other formats. the more export/import options, the better for the users. not all users prefer what we think is the "best" format.

**Reply by guanlisheng** [reply]
Agreed. As tooling, more formats are better. if you already made it and no resource limit, it should be shared to more audience. :)

**Reply by georgeef** [reply]
it is in github. not usable by Windows users.

**Reply by vomikan** [reply]
> it is in github. not usable by Windows users.

Python pandas can read excel files.
What is not a universal solution?

**Reply by georgeef** [reply]
the excel file is of course usable, but the tool ([mmexutil](https://github.com/georgeef/mmexutil)) is in Perl, so not easy to run on Windows. it is possible, but not easy (i didn't try because i don't use Windows).

**Reply by vomikan** [reply]
I'll give you an example to compare the code
You can use the pandas library in Python to convert an Excel file to a CSV file. Here's a simple example:

```python
import pandas as pd

# Read the Excel file
df = pd.read_excel('input.xlsx')

# Write the data to a CSV file
df.to_csv('output.csv', index=False)
```

In this code:

- pd.read_excel('input.xlsx') reads the Excel file 'input.xlsx' into a pandas DataFrame.
- df.to_csv('output.csv', index=False) writes the DataFrame to a CSV file 'output.csv'. The index=False parameter prevents pandas from writing row indices into the CSV file.

Make sure to replace 'input.xlsx' and 'output.csv' with the actual paths to your input Excel file and desired output CSV file.

If you don't have pandas installed, you can install it using pip:

```pip install pandas```


Also, note that this code requires the openpyxl library to be installed if you're working with .xlsx files. You can install it using pip:

```pip install openpyxl```

**Reply by georgeef** [reply]
the conversion between xlsx and csv is ok, it can also be done in LibreOffice. i mean my tool which exports/imports between mmb and xlsx is written in Perl, and it is not easy to run it on Windows.

you can export mmb to csv and then convert to xlsx. but this does not export the whole database, as my tool does. by "whole database" i mean all tables, not only the transactions. same for import, my tool can import all tables.

**Reply by georgeef** [reply]
why export/import to xlsx if you can do the same with csv? in principle it is the same information, as long as you use only primitive xlsx features. but when you have many tables, csv starts becoming inconvenient for the user. in xlsx you can have all tables in one file (as different sheets), in csv you need many files. you can easily open xlsx with Excel/LibreOffice to inspect or process the data. you can do the same with csv, but you need to load multiple files.

this convenience is the only reason i used xlsx as interchange format for my data, and i went into the trouble to program it. read/write in csv (plain text) is of course much easier than read/write in xlsx.

**Comment by georgeef** [comment]
Based on feedback I created proposals for the following new features in MMEX Desktop:
- #6884
- #6885
- #6886
- #6887

Not proposed:
- #6837 (to be closed)
- Print short reports in plain text (not suitable for the MMEX Desktop interface)

Further details can be discussed in the individual proposals.

Let me know if something is missing before I close this discussion.
